File: Source/Run/SPA/src/app/models/investment.ts
export enum AccountType {
  // Fixed-income / cash
  CheckingAccount = 1,
  SavingsBox = 2,
  FixedTerm = 3,
  // Variable-income (renda variável)
  Stock = 4,
  FII = 5,
  ETF = 6,
  InvestmentFund = 7,
  BDR = 8,
  Crypto = 9,
}

export const AccountTypeLabels: Record<AccountType, string> = {
  [AccountType.CheckingAccount]: 'Checking Account',
  [AccountType.SavingsBox]: 'Savings Box',
  [AccountType.FixedTerm]: 'Fixed Term',
  [AccountType.Stock]: 'Stock',
  [AccountType.FII]: 'FII',
  [AccountType.ETF]: 'ETF',
  [AccountType.InvestmentFund]: 'Investment Fund',
  [AccountType.BDR]: 'BDR',
  [AccountType.Crypto]: 'Crypto',
};

const VARIABLE_INCOME_TYPES = new Set<AccountType>([
  AccountType.Stock,
  AccountType.FII,
  AccountType.ETF,
  AccountType.InvestmentFund,
  AccountType.BDR,
  AccountType.Crypto,
]);

export function isVariableIncome(type: AccountType): boolean {
  return VARIABLE_INCOME_TYPES.has(type);
}

export function isFixedTerm(type: AccountType): boolean {
  return type === AccountType.FixedTerm;
}

export function hasDeterministicYield(type: AccountType): boolean {
  return !isVariableIncome(type);
}

export enum Currency {
  BRL = 1,
  USD = 2,
  EUR = 3,
}

export const CurrencyLabels: Record<Currency, string> = {
  [Currency.BRL]: 'BRL',
  [Currency.USD]: 'USD',
  [Currency.EUR]: 'EUR',
};

export interface Investment {
  id: string;
  financialInstitutionId: string;
  institutionName?: string;
  name: string;
  accountType: AccountType;
  currency: Currency;
  amount: number;
  cdiPercentage: number;
  annualRatePercent?: number;
  maturityDate?: string;
  requiresMonthlyMovement: boolean;
  monthlyMovementAmount?: number;
  ticker?: string;
  quantity?: number;
  averagePrice?: number;
  createdDate: string;
  updatedDate: string;
}

export interface InvestmentRequest {
  id?: string;
  financialInstitutionId: string;
  name: string;
  accountType: AccountType;
  currency: Currency;
  amount: number;
  cdiPercentage: number;
  annualRatePercent?: number;
  maturityDate?: string;
  requiresMonthlyMovement: boolean;
  monthlyMovementAmount?: number;
  ticker?: string;
  quantity?: number;
  averagePrice?: number;
}

File: Source/_Core/Contracts/Domain/Investment.cs
using WealthLedger.Contracts.Domain.Enums;

namespace WealthLedger.Contracts.Domain;

public class Investment : BaseEntity
{
    public Guid FinancialInstitutionId { get; set; }
    public string Name { get; set; } = string.Empty;
    public AccountType AccountType { get; set; }
    public Currency Currency { get; set; } = Currency.BRL;
    public decimal Amount { get; set; }
    public decimal CdiPercentage { get; set; }
    public decimal? AnnualRatePercent { get; set; }
    public DateTime? MaturityDate { get; set; }
    public bool RequiresMonthlyMovement { get; set; }
    public decimal? MonthlyMovementAmount { get; set; }

    /// <summary>Ticker / symbol for variable-income (e.g. PETR4, HGLG11, IVVB11, BTC).</summary>
    public string? Ticker { get; set; }

    /// <summary>Number of shares, quotas, or units held (variable-income).</summary>
    public decimal? Quantity { get; set; }

    /// <summary>Average acquisition price per unit (variable-income cost basis).</summary>
    public decimal? AveragePrice { get; set; }
}
